Validate literals in expression analyzer

findepi opened the pull request:
As a result, optimizer can safely assume a `Literal` to represent a
value of given type. Also, queries with invalid literal values should be
guaranteed to fail, regardless of expression pruning.

Fixes https://github.com/trinodb/trino/issues/10719 
Fixes https://github.com/trinodb/trino/issues/10755

kasiafi reviewed (commented):
Please cover also:
- `GenericLiteral` with type `JSON`,
- `IntervalLiteral`

In fact, you could use `LiteralInterpreter` for the validation.

findepi commented:
> In fact, you could use `LiteralInterpreter` for the validation.

I thought about that, i chose not to, because i already have resolved coercion at hand, so `LiteralInterpreter` didn't buy me much.

findepi commented:
> In fact, you could use `LiteralInterpreter` for the validation.

switched over to LiteralInterpreter, as otherwise JSON verification would be cumbersome



> Please cover also:
> 
> * `GenericLiteral` with type `JSON`,
> * `IntervalLiteral`

added

findepi commented on core/trino-main/src/test/java/io/trino/sql/analyzer/TestAnalyzer.java:
```diff
@@ -3250,8 +3250,142 @@ public void testIfInJoinClause()
+        // json
+        assertFails("SELECT JSON ''")
+                .hasErrorCode(INVALID_LITERAL);
+        // TODO https://github.com/trinodb/trino/issues/10724 these should fail
```
https://github.com/trinodb/trino/issues/10724

findepi commented on core/trino-main/src/test/java/io/trino/type/TestDecimalOperators.java:
```diff
@@ -191,10 +192,12 @@ public void testMultiply()
-        assertInvalidFunction("DECIMAL '.12345678901234567890123456789012345678' * DECIMAL '9'", NUMERIC_VALUE_OUT_OF_RANGE);
-        assertInvalidFunction("DECIMAL '12345678901234567890123456789012345678' * DECIMAL '-9'", NUMERIC_VALUE_OUT_OF_RANGE);
-        assertInvalidFunction("DECIMAL '.12345678901234567890123456789012345678' * DECIMAL '-9'", NUMERIC_VALUE_OUT_OF_RANGE);
+        // TODO (https://github.com/trinodb/trino/issues/10732) the expressions here may fail or not, depending whether they are wrapped
+        //   should deterministically fail with NUMERIC_VALUE_OUT_OF_RANGE
```
https://github.com/trinodb/trino/issues/10732

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java:
```diff
@@ -5440,7 +5440,7 @@ public void testTry()
 
         // test try with invalid JSON
-        assertQuery("SELECT JSON_FORMAT(TRY(JSON 'INVALID'))", "SELECT NULL");
+        assertQueryFails("SELECT JSON_FORMAT(TRY(JSON 'INVALID'))", "line 1:24: 'INVALID' is not a valid json literal");
```
This is a behavioral change, but IMO justified.

findepi commented on testing/trino-tests/src/test/java/io/trino/tests/AbstractTestEngineOnlyQueries.java:
```diff
@@ -5455,7 +5455,7 @@ public void testTry()
         assertQuery("SELECT 123 + TRY(ABS(-9223372036854775807 - 1))", "SELECT NULL");
         assertQuery("SELECT JSON_FORMAT(TRY(JSON '[]')) || '123'", "SELECT '[]123'");
-        assertQuery("SELECT JSON_FORMAT(TRY(JSON 'INVALID')) || '123'", "SELECT NULL");
+        assertQueryFails("SELECT JSON_FORMAT(TRY(JSON 'INVALID')) || '123'", "line 1:24: 'INVALID' is not a valid json literal");
```
This is a behavioral change -- see above.

findepi commented:
rebased to resolve conflict with https://github.com/trinodb/trino/pull/10737

findepi commented:
To be rebased after https://github.com/trinodb/trino/pull/10783 is merged.
